internal/resources/gitrepository/spec/spec.go

```diff
@@ -0,0 +1,153 @@
+			branchKey: {
+				Type:        schema.TypeString,
+				Description: "Branch from git to checkout. When branch is given, then that branch from the git repository will be checked out. If the given branch doesn’t exist in the git repository, then adding the git repository will fail. If no branch is given, the `master` branch will be used.",
+				Optional:    true,
+				Default:     "",
+			},
+			tagKey: {
+				Type:        schema.TypeString,
+				Description: "Tag from git to checkout. Takes precedence over branch. When a tag is given, that tag from the git repository will be checked out. If the given tag doesn’t exist in the git repository, then adding the git repository will fail. If both tag and branch are given, tag overrides branch and the branch value will be ignored.",
+				Optional:    true,
+				Default:     "",
+			},
+			semverKey: {
+				Type:        schema.TypeString,
+				Description: "SemVer expression to checkout from git tags. Takes precedence over tag. When semver is given, then the latest tag matching that semver will be checked out from the git repository. If no tag in the git repository matches semver, then adding the git repository will fail. If semver is given, tag and branch will be ignored if they are populated.",
+				Optional:    true,
+				Default:     "",
+			},
+			commitKey: {
+				Type:        schema.TypeString,
+				Description: "Commit SHA to checkout. Takes precedence over all other reference fields. When git_implementation is `GO_GIT`, this can be combined with branch to shallow clone branch in which the commit is expected to exist.",
+				Optional:    true,
+				Default:     "",
+			},
+		},
+	},
+}
+
+func expandRef(data []interface{}) (ref *gitrepositoryclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceFluxcdGitrepositoryReference) {
+	if len(data) == 0 || data[0] == nil {
+		return ref
+	}
+
+	refData, _ := data[0].(map[string]interface{})
+
+	ref = &gitrepositoryclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceFluxcdGitrepositoryReference{}
+
+	if v, ok := refData[branchKey]; ok {
+		helper.SetPrimitiveValue(v, &ref.Branch, branchKey)
+	}
+
+	if v, ok := refData[tagKey]; ok {
+		helper.SetPrimitiveValue(v, &ref.Tag, tagKey)
+	}
+
+	if v, ok := refData[semverKey]; ok {
+		helper.SetPrimitiveValue(v, &ref.Semver, semverKey)
+	}
+
+	if v, ok := refData[commitKey]; ok {
+		helper.SetPrimitiveValue(v, &ref.Commit, commitKey)
+	}
+
+	return ref
+}
+
+func flattenRef(ref *gitrepositoryclustermodel.VmwareTanzuManageV1alpha1ClusterNamespaceFluxcdGitrepositoryReference) (data []interface{}) {
+	if ref == nil {
+		return data
+	}
+
+	flattenRefData := make(map[string]interface{})
+
+	flattenRefData[branchKey] = ref.Branch
+	flattenRefData[tagKey] = ref.Tag
+	flattenRefData[semverKey] = ref.Semver
+	flattenRefData[commitKey] = ref.Commit
+
+	return []interface{}{flattenRefData}
+}
+
+func HasSpecChanged(d *schema.ResourceData) bool {
+	updateRequired := false
+
+	switch {
```

Comment:
check on this logic ? 
You will be setting update required only if all the values have changed

Reply:
I have verified this logic and manually tested all the edge cases after implementing the operation/context handlers. 

This routine returns `true` when one or more values change in the spec due to the `fallthrough`.